socs/agents/hwp_pid/agent.py

```diff
@@ -27,88 +84,116 @@ def __init__(self, agent, ip, port, verbosity):
         self.ip = ip
         self.port = port
         self._verbosity = verbosity > 0
+        self.action_queue = queue.Queue()
 
-        agg_params = {'frame_length': 60}
-        self.agent.register_feed(
-            'hwppid', record=True, agg_params=agg_params)
+        agg_params = {"frame_length": 60}
+        self.agent.register_feed("hwppid", record=True, agg_params=agg_params)
 
-    @ocs_agent.param('auto_acquire', default=False, type=bool)
-    @ocs_agent.param('force', default=False, type=bool)
-    def init_connection(self, session, params):
-        """init_connection(auto_acquire=False, force=False)
+    def _get_data_and_publish(self, pid: pd.PID, session: ocs_agent.OpSession):
+        data = {"timestamp": time.time(), "block_name": "HWPPID", "data": {}}
 
-        **Task** - Initialize connection to PID
-        Controller.
+        current_freq = pid.get_freq()
+        target_freq = pid.get_target()
+        direction = pid.get_direction()
 
-        Parameters:
-            auto_acquire (bool, optional): Default is False. Starts data
-                acquisition after initialization if True.
-            force (bool, optional): Force initialization, even if already
-                initialized. Defaults to False.
+        data["data"]["current_freq"] = current_freq
+        data["data"]["target_freq"] = target_freq
+        data["data"]["direction"] = direction
+
+        session.data.update(
+            {
+                "current_freq": current_freq,
+                "target_freq": target_freq,
+                "direction": direction,
+                "last_updated": time.time(),
+            }
+        )
+        self.agent.publish_to_feed("hwppid", data)
+
+    def _process_actions(self, pid):
+        while not self.action_queue.empty():
+            action = self.action_queue.get()
+            try:
+                self.log.info(f"Running action {action}")
+                res = action.process(pid)
+                action.deferred.callback(res)
+            except Exception as e:
+                self.log.error(f"Error processing action: {action}")
+                action.deferred.errback(e)
```

Comment:
I'm not sure I understand this construction, is it just passing the exception somewhere? I thought `e` had to be a `Callable` to be passed to `.errback()`.